### ISSUE TITLE
Map / Reduce Magnets

### ISSUE THREAD

**Issue by QuincyLarson** [freeCodeCamp/assets#133]
Imagine being able to move magnets around to build an algorithm. I think we can make this happen! We can create a set of magnets that covers some common JavaScript keywords, and allows for people to use arrays and objects, functions and loops.
- [ ] Decide which items to make magnets for
- [ ] Design the magnets
- [ ] Determine cost viability (ideally, I'd like to be able to sell the set for $20 or so), and fit the entire set into a letter-sized envelope so we can use inexpensive international shipping

![](http://www.globalnerdy.com/wordpress/wp-content/uploads/2016/06/map-filter-reduce-in-emoji-1.png)
(maybe we could use images instead of variable names to keep things fun?)

Here are some keywords I think would be good to kick off the discussion:
var
for
while
() =>
i++
<, >, ===, !==
+, -, /, *, %
return
map
filter
reduce

We want to focus on only those that would be heavily used. Perhaps @evaristoc could do a token analysis of our algorithm challenge solutions?

**Comment by QuincyLarson** [comment]
We could actually layer the magnets on top of one another so we could have:

So one magnet could be:
Map([               ],         )

And then you could put icons inside the bracket, along with something like a fire for the function.

This way, with only 20 or 30 magnets, we could have a ton of possible combiations. This would also avoid having awkward parenthesis magnets and such.

**Comment by Wes26** [comment]
I have no idea what the user rights are for emoji considering they are licensed under apple lol. Here are some creative common emoji's that we could possibly use for this asset. http://emojione.com/licensing/  
http://emojione.com/

Wes

**Comment by QuincyLarson** [comment]
@Wes26 wow - these are amazing. As long as we credit the creator, we can use these in commercial projects. I recommend we design the typographic elements using a similar color pallet and curate a set of emojis from this list (so they'll be visually consistent).

My thinking at this point is we keep things simple. Everything should follow this pattern:



map([ 1  2  3 ], 4) => [ 5 6 7 ] 
Where each number would be an emoji. We can leave out the commas, allowing campers to add 1 - 3 emoji in each array. If we use a wide variety of concepts, many of them will serve as both nouns and as verbs. 

For example:

![](http://emojione.com/wp-content/uploads/assets/emojis/2620.svg) can serve as both "death" and "to die"

![](http://emojione.com/wp-content/uploads/assets/emojis/23f3.svg) can serve as both "time" and "to wait"

Some other concepts I think we should include:
![](http://emojione.com/wp-content/uploads/assets/emojis/1f4b0.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f6e0.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f5a5.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f4f1.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f579.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f37b.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f355.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f370.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f37f.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f680.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f3c1.svg)
![](http://emojione.com/wp-content/uploads/assets/emojis/1f308.svg)

Basically, open ended fun imagery that people could re-combine creatively.

**Comment by QuincyLarson** [comment]
I just realized that Filter would require duplicate emoticons. I propose we simplify this to be "Map-Reduce Magnets" with three wide magnets:
`map([             ],      ) =>` and a `[              ]` for the results of map.

`reduce([            ],     ) =>`

We can safely include about 16 emoticons and still be below 1 lbs in shipping weight.
